### DISCUSSION TITLE
Problem with updating a table.

### DISCUSSION THREAD

**Discussion by archsolar** [simolus3/drift#2746] [Q&A] [answered]
Hi when I try to update table: `TodoList` the list doesn't show up any more in a previous page.
Could it be because of `watchTodoItemsInList(..)` which is active?

```
  Stream<List<TodoItem>> watchTodoItemsInList(TodoList todoList) {
    return (select(todoItems)..where((t) => t.todoListId.equals(todoList.id)))
        .watch();
  }
```
Relevant code:
```
  /// watches all todoLists in a given profile. The stream will automatically
  /// emit new items whenever the underlying data changes.
  Stream<List<TodoList>> watchTodoListsInProfile(Profile profile) {
    return (select(todoLists)..where((t) => t.profileId.equals(profile.id)))
        .watch();
  }

```
```
   if (widget._list.percentage != percentage) {
      print("triggered 1");
      // problematic code
      var updated = TodoList(
          id: widget._list.id,
          name: widget._list.name,
          archived: widget._list.archived,
          percentage: widget._list.percentage);

      Global.database.updateTodoList(updated);
    }
```
```
  /// update todoList
  Future updateTodoList(TodoList list) {
    return update(todoLists).replace(list);
  }
```

**Comment by simolus3** [comment]
Is `updateTodoList` doing an update? But I don't understand the full problem yet, does the list just disappear or does a stream stop emitting data?

**Reply by archsolar** [reply]
the stream does not show the updated items. Is that supposed to happen? 🤔

**Reply by archsolar** [reply]
I think I did something wrong here? 

```
  /// update todoList
  Future updateTodoList(TodoList list) {
    return update(todoLists).replace(list);
  }
```

**Reply by archsolar** [reply]
Also when I restart the app the updated `TodoLists` are not shown. but when I inspect the db.sqlite file, things seem to be saved there.

All in all I'm unfamiliar with SQL  and very confused.

**Reply by simolus3** [reply, marked as answer]
> the stream does not show the updated items. Is that supposed to happen? 🤔

No, and the update method looks fine to me as well.

> Also when I restart the app the updated `TodoLists` are not shown. but when I inspect the db.sqlite file, things seem to be saved there.

Maybe something with the query is wrong, are you selecting from the right profile id? You can use `logStatements: true` on the `NativeDatabase` constructor to make drift print the statements it's executing - does running that statement on the `db.sqlite` file show the updated data?

Actually there might be a problem with this

```dart
      // problematic code
      var updated = TodoList(
          id: widget._list.id,
          name: widget._list.name,
          archived: widget._list.archived,
          percentage: widget._list.percentage);
```

It's not setting the profile id, so it will be null after the update - `replace` really does replace the entire row. To only update some columns, use [companion classes](https://drift.simonbinder.eu/docs/dart-api/writes/#updates-and-deletes) also generated by drift.

**Reply by archsolar** [reply]
You are brilliant, thank you!
